### PR TITLE
Modify/Upgrade how the FMA CPP macro works.

### PR DESCRIPTION
**Pull request by KineticTheory**
### Background

* Jayenne uses the FMA macro to select accurate vs fast fused-multiply-add operations.  This allows us to use a high precision software version when hardware FMA are not available.

### Purpose of Pull Request

* [Related to Redmine Issue #1363](https://rtt.lanl.gov/redmine/issues/1363)
* [Related to Redmine Issue #1364](https://rtt.lanl.gov/redmine/issues/1364)

### Description of changes

* Eliminate the `FP_ACCURATE_FMA` mode for machines w/o hardware FMA.  Replace this with the CPP macro `FMA_ACCURATE`.
* For machines with hardware FMA, use the software call if it is faster, otherwise send `a*b+c` to the compiler
* For machines w/o hardware FMA, only use the very slow software call when `FMA_ACCURATE` is requested.
* Provide a special developer toggle `FMA_FIND_DIFFS` that can be set to have every call to FMA compute values using the software call and by direct evaluation of `a*b+c`.  When this mode is active, a report will be printed if the two calculations have a relative difference greater than 1.0e-13.  This is very useful for finding code that is sensitive to round-off errors.

### Status

* Reference:
  * [Pre-Merge Code Review](https://github.com/lanl/Draco/wiki/Style-Guide)
  * [CDash Status](https://rtt.lanl.gov/cdash/index.php?project=Draco&filtercount=1&showfilters=1&field1=buildname&compare1=63&value1=-pr)
* Checks
  * [x] Travis CI checks pass
  * [x] Code coverage does not decrease
  * [x] [Valgrind test passes](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [x] [Toss3 checks pass](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [x] [Trinitite checks pass](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [ ] Code reviewed/approved, sufficient DbC checks, testing, documentation

### CODE REVIEW

File: src/ds++/FMA.hh [hunk 1]
```diff
@@ -7,16 +7,34 @@
...
+ * However, \c fma can be slow (maybe 3x slower).
  *
  * To improve solution consistency across platforms, we can choose to use a
  * software-based infinite precision fma call.
  *
  * \note Intel's code for detecting FMA availability on hardware.
  * https://software.intel.com/en-us/node/405250?language=es&wapkw=avx2+cpuid
+ *
+ * \note The Intel compiler doesn't seem to set \c FP_FAST_FMA. If you know that
+ * your algorithm is sensitive to roundoff, you might to locally disable
```
Comment (clevelam) [inline src/ds++/FMA.hh]: might to locally -> might need to locally

File: src/ds++/FMA.hh [hunk 2]
```diff
@@ -26,41 +44,129 @@
...
+ * Working on LANL systems, it seems that \c FP_FAST_FMA is not set by the Intel
+ * compiler, so Intel will always pick the 2nd option.
+ *
+ * The intent of the \c FMA_ACCURATE version is to pick the more accurate
+ * multiply-plus-add operation. When hardware FMA is available, we just just the
+ * normal 'fast' version.
  */
 #ifdef FP_FAST_FMA
-#define FMA(a, b, c) fma(a, b, c)
+#define FMA(a, b, c) fma((a), (b), (c))
```
Comment (clevelam) [inline src/ds++/FMA.hh]: It seems to me that this ifdef is switchted. Isn't ((a)*(b)+(c)) always the fastest?

Comment (clevelam) [reply src/ds++/FMA.hh]: Also, why do we need FMA and FMA_ACCURATE, it looks like there definitions are always the same in these if blocks.

File: src/ds++/FMA.hh [hunk 3]
```diff
@@ -26,41 +44,129 @@
...
+ * This will be the default for new machines. Both versions should be equally
+ * accurate, choose whatever is faster. \c FP_FAST_FMA is set by the compiler
  *
  * \ref http://en.cppreference.com/w/c/numeric/math/fma
+ *
+ * Working on LANL systems, it seems that \c FP_FAST_FMA is not set by the Intel
+ * compiler, so Intel will always pick the 2nd option.
+ *
+ * The intent of the \c FMA_ACCURATE version is to pick the more accurate
+ * multiply-plus-add operation. When hardware FMA is available, we just just the
```
Comment (clevelam) [inline src/ds++/FMA.hh]: just just -> just

File: src/ds++/FMA.hh [hunk 4]
```diff
@@ -26,41 +44,129 @@
...
+ * that will call an infinite precision fma function so that numerical
+ * reproducibility is enhanced This allows solutions generated on older machines
+ * to match solutions generated on newer machines.
+ *
+ * Operations that are known to be very sensitive to round-off error should use
+ * \c FMA_ACCURATE even though it is much slower (2-3x slower).
+ */
 #define FMA(a, b, c) ((a) * (b) + c)
-#endif
+#define FMA_ACCURATE(a, b, c) fma((a), (b), (c))
```
Comment (clevelam) [inline src/ds++/FMA.hh]: I see, this is where the FMA_ACCURATE and FMA are different.

Comment (clevelam) [reply src/ds++/FMA.hh]: Is this going to bite us if we fail to pickup the native FMA on the platform?

File: src/ds++/FMA.hh (hunk 1 above)
Comment (KineticTheory) [reply src/ds++/FMA.hh]: Thanks for catching this.

File: src/ds++/FMA.hh (hunk 2 above)
Comment (KineticTheory) [reply src/ds++/FMA.hh]: * If `FP_FAST_FMA` is set, then the call to `fma(a,b,c)` is guaranteed to be the fastest computation (yes - faster than `a*b+c` and yes - it is confusing).  You can read the section at http://en.cppreference.com/w/c/numeric/math/fma for more details.
* I want the CPP macro `FMA_ACCURATE` to always be defined.  This allows us to always choose the more _accurate_ version, even when it is slower.  This is required in a few places to allow our tests to pass.  For machines that have hardware FMA, `FMA_ACCURATE == FMA`, but this isn't true for machines w/o hardware FMA.

File: src/ds++/FMA.hh (hunk 3 above)
Comment (KineticTheory) [reply src/ds++/FMA.hh]: fixed.

File: src/ds++/FMA.hh (hunk 4 above)
Comment (KineticTheory) [reply src/ds++/FMA.hh]: * Right
* We will deal with that when we encounter it.  As always, the problem is reproducibility of solutions between platforms and compilers and compiler options.  We could always choose the 'fastest', but then our test setup is worthless.

Comment (KineticTheory) [reply src/ds++/FMA.hh]: I'm sure we will be revisiting this whole issue later this year due to porting to GPU and ARM.

File: src/ds++/FMA.hh [hunk 5]
```diff
@@ -26,41 +44,129 @@
...
+ * multiply-plus-add operation. When hardware FMA is available, we just just the
+ * normal 'fast' version.
  */
 #ifdef FP_FAST_FMA
-#define FMA(a, b, c) fma(a, b, c)
+#define FMA(a, b, c) fma((a), (b), (c))
+#define FMA_ACCURATE(a, b, c) fma((a), (b), (c))
 #else
 #define FMA(a, b, c) ((a) * (b) + c)
+#define FMA_ACURATE(a, b, c) ((a) * (b) + c)
```
Comment (keadyk) [inline src/ds++/FMA.hh]: It seems "accurate" is misspelled here!

Comment (KineticTheory) [reply src/ds++/FMA.hh]: Fixed.  Thanks for catching that one.

**Comment by keadyk** [comment]
Got it! :)

**Pull request merged**